text/0070-disallow-empty-namespace-shadowing.md

```diff
@@ -0,0 +1,166 @@
+types, where you can still refer to the shadowed type using `__cedar`, again as
+explained in [RFC 24].
+Conceptually, primitive and extension types are not declared in the empty
+namespace; they are declared in the `__cedar` namespace, and implicitly imported
+as part of a prelude.
+That is how we could explain (e.g. in the Cedar docs) why the rule is different
+for shadowing entity/common types and shadowing primitive/extension types.
+(Internally, another reason for continuing to allow shadowing
+primitive/extension types, is that we want to be able to introduce new extension
+types without any possibility of breaking existing schemas.)
+
+This RFC does not change anything about [cedar#579].
+We still plan to implement [cedar#579], which allows users to refer to types in
+the empty namespace, in cases where they are not shadowed.
+
+This RFC does apply the new restriction uniformly in both the human and JSON
+schema syntaxes.
+The motivation and situation is the same in both syntaxes for the case of an
+entity typename shadowing an entity typename, and for the case of a common
+typename shadowing a common typename.
+For the case of an entity typename shadowing a common typename or vice versa,
+this RFC still proposes to disallow in both syntaxes, even though these
+situations are not ambiguous in the JSON syntax, which distinguishes between
+entity type and common type references.
+See [this RFC discussion thread](https://github.com/cedar-policy/rfcs/pull/70#discussion_r1659120108).
+
+## Drawbacks
+
+### Drawback 1
+
+The motivation section for [RFC 64] claimed that
+> Users justifiably expect that they can refer to any declared typename (in any
+> namespace) from any position in the schema, by writing a fully qualified name.
+> The fact that [RFC 24] did not provide a way to refer to typenames in the empty
+> namespace by using some kind of fully qualified name seems like an oversight,
+> and should be corrected for consistency.
+
+This RFC does not address that motivation, and effectively counterclaims that
+users don't actually expect this.
+
+### Drawback 2
+
+If [RFC 69] or something similar are eventually accepted, the error in this RFC
+may not be so easily avoided, if the empty-namespace definition is part of a
+third-party library being imported by the user.
+In that case, the user couldn't easily rename the empty-namespace type or move
+it into a nonempty namespace.
+
+To address this drawback, we should guide library authors to not declare library
+types in the empty namespace (and instead, e.g., declare them in a namespace
+named after the library).
+We could even codify that rule, requiring that all libraries have a name and all
+declarations in the library live in the corresponding namespace, or alternately,
+implicitly prefixing all library declarations with the library name as a
+toplevel namespace.
+Many existing package systems have similar rules or mechanisms.
+
+## Alternatives
```

Comment:
A note: This RFC does not say anything about shadowing for _action_ declarations.  In fact, for actions (unlike for entity/common types) we already have syntax that allows you to explicitly reference empty-namespace declarations:

```
// refers to `action Group` in the current namespace,
// or, after #579 is fixed, `action Group` in the empty namespace
// if there is no `action Group` in the current namespace
action A in [Group];

// Same as the first example; this syntax allows for action IDs that 
// contain spaces or special characters etc.
action A in ["Group"];

// refers to `action Group` in another specified namespace
action A in [OtherNS::Action::"Group"];

// refers to `action Group` in the empty namespace specifically, 
// even if there is also an `action Group` in the current namespace
action A in [Action::"Group"];
```

At least, this is my current understanding. Happy to be corrected if I'm wrong.

Reply:
Based on cedar-policy/cedar#150 I'd expect that the last option would also be qualified with the current namespace

Reply:
Actually, from looking at the code, I suspect the fourth example is also equivalent to the first two examples.  (Not sure if that's what we want.  Also not sure if changing it is worth a breaking change though.)

Reply:
oh didn't see your comment @john-h-kastner-aws when I wrote that.  So we're in agreement.

In that case, the situation for actions is more similar to the situation for entity/common types.  Should we disallow action shadowing as well?  (I.e., disallow declaring `action Group` in both the empty namespace and a non-empty namespace?)

Reply:
Updated the RFC to disallow this shadowing for action names as well.